### PR TITLE
whole app

### PR DESCRIPTION
**Pull request by serejabogdan**
added redux, saga, react-router-dom, connected-react-router, axios, redux-form
made pages - sign-in form and user information
added app initializing

### CODE REVIEW

File: src/components/input-field/index.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  input: PropTypes.object,
+  type: PropTypes.string,
+  id: PropTypes.string,
+  label: PropTypes.string,
+  meta: PropTypes.shape({
+    touched: PropTypes.bool,
+    warning: PropTypes.string,
+    error: PropTypes.string
+  }),
+  disabled: PropTypes.bool
```
Comment (ichabanivan) [inline src/components/input-field/index.js]: add isRequired props

File: src/components/preloader/index.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import React from 'react';
```
Comment (ichabanivan) [inline src/components/preloader/index.js]: Use reactstrap Spinner

File: src/components/user-data/index.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+import { ListGroup, ListGroupItem } from 'reactstrap';
+import PropTypes from 'prop-types';
+
+function UserData ({ user }) {
+  const { id, name, coverImage } = user;
+  const { url } = coverImage;
+  return (
+    <div style={{ maxWidth: '700px', margin: '0 auto' }}>
+      <ListGroup>
+        <ListGroupItem>id: { id }</ListGroupItem>
```
Comment (ichabanivan) [inline src/components/user-data/index.js]:
```suggestion
        <ListGroupItem>id: { user.id }</ListGroupItem>
```

File: src/index.css [hunk 4]
```diff
@@ -0,0 +1,3 @@
+body {
+    text-align: center;
```
Comment (ichabanivan) [inline src/index.css]:
```suggestion
```

File: src/index.js [hunk 5]
```diff
@@ -1,10 +1,19 @@
 import React from 'react';
```
Comment (ichabanivan) [inline src/index.js:2]: // outsource dependencies

File: src/index.js [hunk 6]
```diff
@@ -1,10 +1,19 @@
...
 import ReactDOM from 'react-dom';
-import App from './components/App/App';
+
+import './index.css';
```
Comment (ichabanivan) [inline src/index.js]: // local dependencies

File: src/pages/index.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
+import React, { useEffect } from 'react';
```
Comment (ichabanivan) [inline src/pages/index.js:2]: 🌲

File: src/pages/index.js [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+import { Redirect, Route, Switch } from 'react-router-dom';
+import Public from './public';
+import Private from './private';
+import { useDispatch, useSelector } from 'react-redux';
+import { appInitializing } from './reducer';
+import Preloader from '../components/preloader';
+
+function Pages () {
+  const { initialized } = useSelector(state => state.root.pagesInitialize);
```
Comment (ichabanivan) [inline src/pages/index.js]: const selector = state => state.root.pagesInitialize

File: src/pages/index.js [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+    dispatch(appInitializing(accessToken));
+  }, [accessToken, dispatch]);
+
+  return (
+    <>
+      { initialized
+        ? <Switch>
+          <Route path="/public" component={Public}/>
+          <Route path="/private" component={Private}/>
+          <Route exact path="/" render={() => <Redirect to="/public"/>}/>
```
Comment (ichabanivan) [inline src/pages/index.js]:
```suggestion
          <Redirect to="/public"/>
```

File: src/pages/private/user/index.js [hunk 10]
```diff
@@ -0,0 +1,17 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import UserData from '../../../components/user-data';
+
+export default function User ({ user }) {
+  return (
```
Comment (ichabanivan) [inline src/pages/private/user/index.js]:
```suggestion
  return 
```

File: src/pages/public/index.js [hunk 11]
```diff
@@ -0,0 +1,16 @@
+import React from 'react';
+import { Redirect, Route, Switch } from 'react-router-dom';
+import PublicForm from './sign-in';
+import SignUp from './sign-up';
+
+function Public () {
+  return (
+    <Switch>
+      <Route path="/public/sign-in" component={PublicForm}/>
```
Comment (ichabanivan) [inline src/pages/public/index.js]: routes as constants

File: src/pages/public/sign-in/index.js [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+import 'bootstrap/dist/css/bootstrap.min.css';
+
+import { connect, useDispatch, useSelector } from 'react-redux';
+
+import { reduxForm, Field } from 'redux-form';
+import InputField from '../../../components/input-field';
+import { asyncSetToken } from './reducer';
+
+function PublicForm () {
+  const { initialValues: { client } }
```
Comment (ichabanivan) [inline src/pages/public/sign-in/index.js]:
```suggestion
  const { initialValues }
```

File: src/pages/public/sign-in/index.js [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+  return (
+    <div className="public-form">
+      <h1>Public form</h1>
+      <UserFormRedux onSubmit={submit}/>
+    </div>
+  );
+}
+
+function UserForm (props) {
+  const { disabled } = useSelector(state => state.root.pagesInitialize);
```
Comment (ichabanivan) [inline src/pages/public/sign-in/index.js]: use redux store for sign-in page

File: src/pages/public/sign-in/index.js [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+  return (
+    <Form onSubmit={props.handleSubmit}>
+      <FormGroup>
```
Comment (ichabanivan) [inline src/pages/public/sign-in/index.js]: InputField -> FormGroup``

File: src/pages/public/sign-in/reducer.js [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+
+const loginInitialState = {
+  initialValues: {
+    client: 'admin_application'
+  },
+  accessToken: localStorage.getItem('token') || '',
+};
+
+function signInReducer (state = loginInitialState, action) {
+  switch (action.type) {
```
Comment (ichabanivan) [inline src/pages/public/sign-in/reducer.js]: Create META type

File: src/pages/public/sign-in/saga.js [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+  yield put(asyncSetUser(data.accessToken));
+}
+
+function * signinWorker (action) {
+  try {
+    const { data } = yield call(getUserData, action.payload);
+    yield put(setUser(data));
+    yield put(push('/private/user'));
+  } catch (e) {
+    const UNAUTHORIZED = 401;
```
Comment (ichabanivan) [inline src/pages/public/sign-in/saga.js]:
```suggestion
```

File: src/pages/public/sign-in/sign-in.css [hunk 17]
```diff
@@ -0,0 +1,7 @@
+.public-form {
+    min-height: 100vh;
+    display: flex;
```
Comment (ichabanivan) [inline src/pages/public/sign-in/sign-in.css]:
```suggestion
```

File: src/pages/reducer.js [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+
+export const SET_USER = 'SET_USER';
+export const ASYNC_SET_USER = 'ASYNC_SET_USER';
+export const SET_INITIALIZED = 'SET_INITIALIZED';
+export const APP_INITIALIZING = 'APP_INITIALIZING';
+export const SET_DISABLED = 'SET_DISABLED';
+
+const initialState = {
+  initialized: false,
+  errorMessage: 'Something wrong',
```
Comment (ichabanivan) [inline src/pages/reducer.js]:
```suggestion
  errorMessage: '',
```

File: src/pages/saga.js [hunk 19]
```diff
@@ -0,0 +1,18 @@
+import { fork, put, takeEvery } from 'redux-saga/effects';
+import publicSaga from './public/saga';
+import { APP_INITIALIZING, asyncSetUser } from './reducer';
+import { delay } from '../utils/API';
```
Comment (ichabanivan) [inline src/pages/saga.js]: delay  from 'redux-saga/effects'

File: src/utils/API.js [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+    data: { ...payload }
+  });
+}
+
+export function getUserData (token) {
+  return API({
+    method: 'GET',
+    url: 'auth/users/me',
+    headers: {
+      Authorization: `Bearer ${token}`
```
Comment (ichabanivan) [inline src/utils/API.js]: https://github.com/axios/axios#config-defaults

File: src/utils/form-validators.js [hunk 21]
```diff
@@ -0,0 +1,28 @@
+export function required (value) {
+  if (value) {
+    // eslint-disable-next-line no-undefined
+    return undefined;
```
Comment (ichabanivan) [inline src/utils/form-validators.js]:
```suggestion
    return null;
```

File: src/utils/form-validators.js [hunk 22]
```diff
@@ -0,0 +1,28 @@
...
+  return function (value) {
+    if (value && value.length > maxCharacters) {
+      return `Must be ${maxCharacters} characters or less`;
+    }
+    // eslint-disable-next-line no-undefined
+    return undefined;
+  };
+}
+
+export const maxLength15 = maxLengthCreator(15);
```
Comment (ichabanivan) [inline src/utils/form-validators.js]:
```suggestion
```

**Review by ichabanivan — CHANGES_REQUESTED** [review]

File: src/components/input-field/index.js [hunk 23]
```diff
@@ -0,0 +1,41 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { FormGroup, Input, Label } from 'reactstrap';
+
+function InputField ({ input, type, id, label, meta, disabled }) {
+  return (
```
Comment (ichabanivan) [inline src/components/input-field/index.js]:
```suggestion
  return <FormGroup>
```

File: src/components/input-field/index.js [hunk 24]
```diff
@@ -0,0 +1,41 @@
...
+  },
+  disabled: false
+};
+
+// TODO: CHECK propTypes and jsx
+
+InputField.propTypes = {
+  input: PropTypes.object.isRequired,
+  type: PropTypes.string.isRequired,
+  id: PropTypes.string,
```
Comment (ichabanivan) [inline src/components/input-field/index.js]:
```suggestion
  name: PropTypes.string.isRequired,
```

File: src/components/preloader/index.js [hunk 25]
```diff
@@ -0,0 +1,8 @@
...
+import { Spinner } from 'reactstrap';
+
+function Preloader () {
+  return <Spinner color="primary" />;
```
Comment (ichabanivan) [inline src/components/preloader/index.js]: Use it directly in the code

File: src/components/user-data/index.js [hunk 26]
```diff
@@ -0,0 +1,35 @@
...
+import PropTypes from 'prop-types';
+import { ListGroup, ListGroupItem } from 'reactstrap';
+
+function UserData ({ user }) {
+  return (
+    <div style={{ maxWidth: '700px', margin: '0 auto' }}>
+      <ListGroup>
+        <ListGroupItem>id: { user.id }</ListGroupItem>
+        <ListGroupItem>name: { user.name }</ListGroupItem>
+        <ListGroupItem><img src={user.coverImage.url} alt=""/></ListGroupItem>
```
Comment (ichabanivan) [inline src/components/user-data/index.js]: alt

File: src/components/user-data/index.js [hunk 27]
```diff
@@ -0,0 +1,35 @@
...
+};
+
+UserData.propTypes = {
+  user: PropTypes.shape({
+    id: PropTypes.number.isRequired,
+    name: PropTypes.string.isRequired,
+    coverImage: PropTypes.shape({
+      url: PropTypes.string
+    })
+  })
```
Comment (ichabanivan) [inline src/components/user-data/index.js]: required

File: src/pages/index.js [hunk 28]
```diff
@@ -0,0 +1,33 @@
+import React, { useEffect } from 'react';
+import { useDispatch, useSelector } from 'react-redux';
+import { Redirect, Route, Switch } from 'react-router-dom';
+
+import Public from './public';
+import Private from './private';
+import Preloader from '../components/preloader';
+import { PRIVATE, PUBLIC } from '../utils/constants';
```
Comment (ichabanivan) [inline src/pages/index.js]:
```suggestion
import { PRIVATE, PUBLIC } from '../constants';
```

File: src/pages/reducer.js [hunk 29]
```diff
@@ -0,0 +1,42 @@
...
+    default:
+      return state;
+  }
+}
+
+export function getAccessToken () {
+  return state => state.root.pagesData.accessToken;
+}
+
+export function getPagesData () {
```
Comment (ichabanivan) [inline src/pages/reducer.js]:
```suggestion
export function getPagesData (state) {
```

File: src/pages/reducer.js [hunk 30]
```diff
@@ -0,0 +1,42 @@
...
+  return state => state.root.pagesData;
```
Comment (ichabanivan) [inline src/pages/reducer.js]:
```suggestion
  return state.root.pagesData;
```

File: src/pages/private/index.js [hunk 31]
```diff
@@ -0,0 +1,27 @@
...
+import User from './user';
+import Homepage from './homepage';
+import { getPagesData } from '../reducer';
+import { getPrivateState } from './reducer';
+import Preloader from '../../components/preloader';
+import PrivateRoute from '../../utils/private-route';
+import { PRIVATE_HOMEPAGE, PRIVATE_USER } from '../../utils/constants';
+
+function Private () {
+  const privateState = useSelector(getPrivateState());
```
Comment (ichabanivan) [inline src/pages/private/index.js]:
```suggestion
  const { initialized } = useSelector(getPrivateState());
```

File: src/pages/private/reducer.js [hunk 32]
```diff
@@ -0,0 +1,21 @@
+export const PRIVATE_SAGA_VALID_TOKEN = 'PRIVATE_SAGA_VALID_TOKEN';
+
+export const PRIVATE_META = 'PRIVATE_META';
```
Comment (ichabanivan) [inline src/pages/private/reducer.js]:
```suggestion
export const META = '@private/META';
```

File: src/pages/private/saga.js [hunk 33]
```diff
@@ -0,0 +1,29 @@
+import { call, delay, fork, put, takeEvery } from 'redux-saga/effects';
+import { push } from 'connected-react-router';
+
+import { PAGES_META } from '../reducer';
+import { getUserData } from '../../utils/API';
+import { PRIVATE_META, PRIVATE_SAGA_VALID_TOKEN } from './reducer';
+import { PRIVATE_USER, PUBLIC_SIGN_IN } from '../../utils/constants';
```
Comment (ichabanivan) [inline src/pages/private/saga.js]:
```suggestion
import { PRIVATE_USER, PUBLIC_SIGN_IN } from '../../constants/routes.js';
```

File: src/pages/private/saga.js [hunk 34]
```diff
@@ -0,0 +1,29 @@
...
+
+function * gettingUserDataWorker (action) {
```
Comment (ichabanivan) [inline src/pages/private/saga.js]:
```suggestion
function * gettingUserDataWorker ({ type, payload }) {
```

File: src/pages/private/saga.js [hunk 35]
```diff
@@ -0,0 +1,29 @@
...
+
+  yield delay(500);
+  try {
+    const response = yield call(getUserData, action.payload);
+    yield put({ type: PAGES_META, payload: { user: response.data } });
+    yield put({ type: PRIVATE_META, payload: { initialized: true } });
+  } catch (error) {
+    yield put(push(PUBLIC_SIGN_IN));
+  }
+}
```
Comment (zanusilker) [inline src/pages/private/saga.js:20]:
```suggestion
yield put({ type: PRIVATE_META, payload: { initialized: true } });
}
```

File: src/pages/private/user/index.js [hunk 36]
```diff
@@ -0,0 +1,11 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import UserData from '../../../components/user-data';
+
+export default function User ({ user }) {
+  return <UserData user={user}/>
```
Comment (ichabanivan) [inline src/pages/private/user/index.js]: why it is another component?

File: src/pages/public/index.js [hunk 37]
```diff
@@ -0,0 +1,18 @@
...
+
+import SignUp from './sign-up';
+import PublicForm from './sign-in';
+import { PUBLIC_SIGN_IN } from '../../utils/constants';
+
+function Public () {
+  return (
+    <Switch>
+      <Route path={PUBLIC_SIGN_IN} component={PublicForm}/>
+      <Route path={PUBLIC_SIGN_IN} component={SignUp}/>
```
Comment (ichabanivan) [inline src/pages/public/index.js]:
```suggestion
      <Route path={PUBLIC_SIGN_UP} component={SignUp}/>
```

File: src/pages/public/index.js [hunk 38]
```diff
@@ -0,0 +1,18 @@
...
+import { Redirect, Route, Switch } from 'react-router-dom';
+
+import SignUp from './sign-up';
+import PublicForm from './sign-in';
+import { PUBLIC_SIGN_IN } from '../../utils/constants';
+
+function Public () {
+  return (
+    <Switch>
+      <Route path={PUBLIC_SIGN_IN} component={PublicForm}/>
```
Comment (ichabanivan) [inline src/pages/public/index.js]:
```suggestion
      <Route path={PUBLIC_SIGN_IN} component={SignIn}/>
```

File: src/pages/public/reducer.js [hunk 39]
```diff
@@ -0,0 +1,10 @@
+import { combineReducers } from 'redux';
+
+import signInReducer from './sign-in/reducer';
+import signUpReducer from './sign-up/reducer';
+
+
```
Comment (ichabanivan) [inline src/pages/public/reducer.js]:
```suggestion
```

File: src/pages/public/saga.js [hunk 40]
```diff
@@ -0,0 +1,34 @@
...
+import { PAGES_META } from '../reducer';
+import { publicAPI } from '../../utils/API';
+import { SAGA_SET_TOKEN } from './sign-in/reducer';
+import { PRIVATE_SAGA_VALID_TOKEN } from '../private/reducer';
+
+export function getTokenFromApi (payload) {
+  return publicAPI({
+    method: 'POST',
+    url: 'auth/token',
+    data: { ...payload }
```
Comment (ichabanivan) [inline src/pages/public/saga.js]:
```suggestion
    data: payload
```

File: src/pages/public/saga.js [hunk 41]
```diff
@@ -0,0 +1,34 @@
...
+  });
+}
+
+function * setTokenWorker (action) {
+  try {
+    const response = yield call(getTokenFromApi, action.payload);
+    yield call([localStorage, localStorage.setItem], 'token', response.data.accessToken);
```
Comment (ichabanivan) [inline src/pages/public/saga.js]:
```suggestion
    yield call(localStorage.setItem, 'token', response.data.accessToken);
```

File: src/pages/public/saga.js [hunk 42]
```diff
@@ -0,0 +1,34 @@
+import { call, fork, put, takeEvery } from 'redux-saga/effects';
+
+import { PAGES_META } from '../reducer';
+import { publicAPI } from '../../utils/API';
+import { SAGA_SET_TOKEN } from './sign-in/reducer';
+import { PRIVATE_SAGA_VALID_TOKEN } from '../private/reducer';
+
+export function getTokenFromApi (payload) {
```
Comment (ichabanivan) [inline src/pages/public/saga.js]: move too sign-in page

File: src/pages/public/sign-in/index.js [hunk 43]
```diff
@@ -0,0 +1,44 @@
...
+import { getClient, SAGA_SET_TOKEN } from './reducer';
+import InputField from '../../../components/input-field';
+
+function SignIn () {
+  const client = useSelector(getClient());
+  const dispatch = useDispatch();
+
+  function submit (values) {
+    const submitData = { client, ...values };
+    dispatch({ type: SAGA_SET_TOKEN, payload: submitData });
```
Comment (ichabanivan) [inline src/pages/public/sign-in/index.js]:
```suggestion
    dispatch({ type: SAGA_SET_TOKEN, payload: { client, ...values } });
```

File: src/pages/public/sign-in/index.js [hunk 44]
```diff
@@ -0,0 +1,44 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { Button, Form } from 'reactstrap';
+import { reduxForm, Field } from 'redux-form';
+import 'bootstrap/dist/css/bootstrap.min.css';
```
Comment (ichabanivan) [inline src/pages/public/sign-in/index.js]: // styles

File: src/pages/public/sign-in/index.js [hunk 45]
```diff
@@ -0,0 +1,44 @@
...
+
+  function submit (values) {
+    const submitData = { client, ...values };
+    dispatch({ type: SAGA_SET_TOKEN, payload: submitData });
+  }
+
+  return (
+    <div style={{ minHeight: '100vh' }} className="d-flex flex-column justify-content-center align-items-center">
+      <h1>Public form</h1>
+      <UserFormRedux onSubmit={submit}/>
```
Comment (ichabanivan) [inline src/pages/public/sign-in/index.js]: why does it another component?

File: src/pages/public/sign-in/reducer.js [hunk 46]
```diff
@@ -0,0 +1,20 @@
+export const SAGA_SET_TOKEN = 'SAGA_SET_TOKEN';
+
+const loginInitialState = {
```
Comment (ichabanivan) [inline src/pages/public/sign-in/reducer.js]:
```suggestion
const initial = {
```

File: src/pages/public/sign-in/reducer.js [hunk 47]
```diff
@@ -0,0 +1,20 @@
...
+};
+
+function signInReducer (state = loginInitialState, action) {
+  switch (action.type) {
+    default:
+      return state;
+  }
+}
+
+export function getClient () {
```
Comment (ichabanivan) [inline src/pages/public/sign-in/reducer.js]:
```suggestion
export function selector () {
```

File: src/pages/public/sign-in/reducer.js [hunk 48]
```diff
@@ -0,0 +1,20 @@
...
+  return state => state.root.pages.public.signin.initialValues.client;
```
Comment (ichabanivan) [inline src/pages/public/sign-in/reducer.js]:
```suggestion
  return state => state.root.pages.public.signin
```

File: src/pages/reducer.js [hunk 49]
```diff
@@ -0,0 +1,42 @@
...
+
+import publicReducer from './public/reducer';
+import privateReducer from './private/reducer';
+
+export const PAGES_META = 'PAGES_META';
+export const APP_INITIALIZING = 'APP_INITIALIZING';
+
+const initialState = {
+  user: '',
+  accessToken: localStorage.getItem('token') || '',
```
Comment (ichabanivan) [inline src/pages/reducer.js]:
```suggestion
  token: localStorage.getItem('token') || '',
```

File: src/pages/saga.js [hunk 50]
```diff
@@ -0,0 +1,30 @@
...
+import privateSaga from './private/saga';
+import { getUserData } from '../utils/API';
+import { push } from 'connected-react-router';
+import { PUBLIC_SIGN_IN } from '../utils/constants';
+import { APP_INITIALIZING, PAGES_META } from './reducer';
+import { PRIVATE_SAGA_VALID_TOKEN } from './private/reducer';
+
+function * appInitializeWorker (action) {
+  yield delay(500);
+  try {
```
Comment (zanusilker) [inline src/pages/saga.js:13]: get token from localstorage

File: src/utils/API.js [hunk 51]
```diff
@@ -0,0 +1,31 @@
...
+  withCredentials: false,
+  headers: {
+    'Cache-Control': 'no-cache',
+    'Content-Type': 'application/json',
+  }
+});
+
+export function getUserData (token) {
+  if (token) {
+    privateAPI.defaults.headers.common.Authorization = `Bearer ${token}`;
```
Comment (ichabanivan) [inline src/utils/API.js]: initialize on pages saga

**Review by ichabanivan — CHANGES_REQUESTED** [review]

File: src/pages/index.js [hunk 52]
```diff
@@ -0,0 +1,33 @@
+// outsource dependencies
+import React, { useEffect } from 'react';
+import { useDispatch, useSelector } from 'react-redux';
+import { Redirect, Route, Switch } from 'react-router-dom';
+
+// local dependencies
+import Public from './public';
+import Private from './private';
+import { Spinner } from 'reactstrap';
```
Comment (ichabanivan) [inline src/pages/index.js]: // outsource dependencies

File: src/pages/index.js [hunk 53]
```diff
@@ -0,0 +1,33 @@
...
+import { PRIVATE, PUBLIC } from '../constants/routes';
+import { selector, PAGES } from './reducer';
```
Comment (ichabanivan) [inline src/pages/index.js]: 🌲

File: src/pages/index.js [hunk 54]
```diff
@@ -0,0 +1,33 @@
...
+
+function Pages () {
+  const { initialized, token } = useSelector(selector);
+  const dispatch = useDispatch();
+
+  useEffect(() => {
+    dispatch({ type: PAGES.INITIALIZE, payload: token });
+    return () => {
+      dispatch({ type: 'STOP_REFRESHING_TOKEN' });
```
Comment (ichabanivan) [inline src/pages/index.js]: use constants

File: src/pages/index.js [hunk 55]
```diff
@@ -0,0 +1,33 @@
...
+    };
+  }, [dispatch, token]);
+  return <>
```
Comment (ichabanivan) [inline src/pages/index.js]:
```suggestion
  return 
```

File: src/pages/private/index.js [hunk 56]
```diff
@@ -0,0 +1,29 @@
+// outsource dependencies
+import React from 'react';
+import { Switch } from 'react-router-dom';
+import { useSelector } from 'react-redux';
+
+// local dependencies
+import User from './user';
+import Homepage from './homepage';
+import { Spinner } from 'reactstrap';
```
Comment (ichabanivan) [inline src/pages/private/index.js]: ?

File: src/pages/private/index.js [hunk 57]
```diff
@@ -0,0 +1,29 @@
...
+import PrivateRoute from '../../utils/private-route';
+import { PRIVATE_HOMEPAGE, PRIVATE_USER } from '../../constants/routes';
+
+function Private () {
+  const root = useSelector(selector);
+
+  return <>
+    {
+      root.pages.private.initialized ? <Switch>
+        <PrivateRoute path={PRIVATE_USER} isUserAuthed={!!root.pagesData.user}
```
Comment (ichabanivan) [inline src/pages/private/index.js]:
```suggestion
        <PrivateRoute 
        path={PRIVATE_USER} 
        isUserAuthed={!!root.pagesData.user}
```

File: src/pages/private/index.js [hunk 58]
```diff
@@ -0,0 +1,29 @@
+// outsource dependencies
+import React from 'react';
+import { Switch } from 'react-router-dom';
+import { useSelector } from 'react-redux';
+
+// local dependencies
+import User from './user';
+import Homepage from './homepage';
+import { Spinner } from 'reactstrap';
+import { selector } from '../../redux/reducer';
```
Comment (ichabanivan) [inline src/pages/private/index.js]:
```suggestion
import { selector as privateSelector } from '../../redux/reducer';
```

File: src/pages/private/reducer.js [hunk 59]
```diff
@@ -0,0 +1,28 @@
+export const PRIVATE = (function (prefix) {
+  return {
+    // simple actions
+    META: `${prefix}META`,
+    CLEAR: `${prefix}CLEAR`,
+    // complex actions
+    VALID_TOKEN: `${prefix}VALID_TOKEN`
```
Comment (ichabanivan) [inline src/pages/private/reducer.js]: where do you use it?

File: src/pages/private/saga.js [hunk 60]
```diff
@@ -0,0 +1,54 @@
...
+import { call, delay, fork, put, takeEvery } from 'redux-saga/effects';
+
+// local dependencies
+import { PAGES } from '../reducer';
+import { PRIVATE } from './reducer';
+import { SAGA_LOGOUT } from './user/reducer';
+import { TOKEN } from '../../constants/local-storage';
+import { PRIVATE_USER, PUBLIC_SIGN_IN } from '../../constants/routes';
+import { getLocalStorage, removeLocalStorage } from '../../utils/local-storage';
+import { addAuthorizationHeader, getUserData, privateAPI } from '../../utils/API';
```
Comment (ichabanivan) [inline src/pages/private/saga.js:11]: extract user request to this file

File: src/pages/private/user/index.js [hunk 61]
```diff
@@ -0,0 +1,34 @@
+// outsource dependencies
+import React from 'react';
+import PropTypes from 'prop-types';
+import { SAGA_LOGOUT } from './reducer';
```
Comment (ichabanivan) [inline src/pages/private/user/index.js]: ?

File: src/pages/private/user/index.js [hunk 62]
```diff
@@ -0,0 +1,34 @@
...
+import { useDispatch } from 'react-redux';
+import { Button, ListGroup, ListGroupItem } from 'reactstrap';
+
+function User ({ user }) {
+  const dispatch = useDispatch();
+
+  function logOut () {
+    dispatch({ type: SAGA_LOGOUT });
```
Comment (ichabanivan) [inline src/pages/private/user/index.js]:
```suggestion
    dispatch({ type: TYPE.SAGA_LOGOUT });
```

File: src/pages/private/user/index.js [hunk 63]
```diff
@@ -0,0 +1,34 @@
...
+  }
+  return <div style={{ maxWidth: '700px', margin: '0 auto' }}>
+    <ListGroup>
+      <ListGroupItem>id: { user.id }</ListGroupItem>
+      <ListGroupItem>name: { user.name }</ListGroupItem>
+      <ListGroupItem><img src={user.coverImage.url} alt="Here is should be user"/></ListGroupItem>
```
Comment (ichabanivan) [inline src/pages/private/user/index.js]:
```suggestion
      <ListGroupItem>
      <img src={user.coverImage.url} alt="Here is should be user"/>
      </ListGroupItem>
```

File: src/pages/private/user/reducer.js [hunk 64]
```diff
@@ -0,0 +1,5 @@
+export const SAGA_LOGOUT = 'SAGA_LOGOUT';
```
Comment (ichabanivan) [inline src/pages/private/user/reducer.js]: PRIVATE

File: src/pages/index.js [hunk 65]
```diff
@@ -0,0 +1,31 @@
...
+    return () => {
+      dispatch({ type: PAGES.STOP_REFRESHING_TOKEN });
+    };
+  }, [dispatch, token]);
+  return initialized
+    ? <Switch>
+      <Route path={PUBLIC} component={Public}/>
+      <Route path={PRIVATE} component={Private}/>
+      <Redirect to={PUBLIC}/>
+    </Switch> : <div><Spinner color="primary" /> pages</div>;
```
Comment (ichabanivan) [inline src/pages/index.js]:
```suggestion
    </Switch> 
    : <div>
    <Spinner color="primary" />
     pages
     </div>;
```

File: src/pages/private/index.js [hunk 66]
```diff
@@ -0,0 +1,34 @@
...
+  const { user } = useSelector(pagesSelector);
+  const { initialized } = useSelector(privateSelector);
+
+  return initialized
+    ? <Switch>
+      <PrivateRoute
+        path={PRIVATE_USER}
+        isUserAuthed={!!user}
+        user={user}
+        component={User}/>
```
Comment (ichabanivan) [inline src/pages/private/index.js]:
```suggestion
        component={User}
        />
```

File: src/pages/private/index.js [hunk 67]
```diff
@@ -0,0 +1,34 @@
...
+function Private () {
+  const { user } = useSelector(pagesSelector);
+  const { initialized } = useSelector(privateSelector);
+
+  return initialized
+    ? <Switch>
+      <PrivateRoute
+        path={PRIVATE_USER}
+        isUserAuthed={!!user}
+        user={user}
```
Comment (ichabanivan) [inline src/pages/private/index.js]: 🌲

File: src/pages/public/saga.js [hunk 68]
```diff
@@ -0,0 +1,14 @@
+// outsource dependencies
+import { fork } from 'redux-saga/effects';
+
+// local dependencies
+import signUpSaga from './sign-up/saga';
+import signInSaga from './sign-in/saga';
+
+function * publicPagesWatcher () {}
```
Comment (ichabanivan) [inline src/pages/public/saga.js]:
```suggestion
```

File: src/pages/public/saga.js [hunk 69]
```diff
@@ -0,0 +1,14 @@
...
+
+export default function * publicSaga () {
+  yield fork(publicPagesWatcher);
```
Comment (ichabanivan) [inline src/pages/public/saga.js]:
```suggestion
```

File: src/pages/public/sign-in/index.js [hunk 70]
```diff
@@ -0,0 +1,55 @@
...
+
+  return errors;
+}
+
+function SignIn () {
+  const signIn = useSelector(selector);
+  const dispatch = useDispatch();
+
+  function submit (values) {
+    dispatch({ type: SIGN_IN.SET_TOKEN, payload: { client: signIn.initialValues.client, ...values } });
```
Comment (ichabanivan) [inline src/pages/public/sign-in/index.js]:
```suggestion
    dispatch({ type: TYPE.SIGN_IN, payload: values });
```

File: src/pages/public/sign-in/index.js [hunk 71]
```diff
@@ -0,0 +1,55 @@
...
+  }
+  return <div style={{ minHeight: '100vh' }} className="d-flex flex-column justify-content-center align-items-center">
+    <h1>Public form</h1>
+    <ReduxForm form="SignInForm" onSubmit={submit} validate={formValidation}>
```
Comment (ichabanivan) [inline src/pages/public/sign-in/index.js]:
```suggestion
    <ReduxForm form="SignInForm" onSubmit={submit} validate={formValidation} initialValues={signIn.data}>
```

File: src/pages/public/sign-in/reducer.js [hunk 72]
```diff
@@ -0,0 +1,32 @@
...
+    //complex actions
+    SET_TOKEN: `${prefix}SET_TOKEN`,
+  };
+})('@sign-in/');
+
+const initial = {
+  initialValues: {
+    client: 'admin_application'
+  },
+  disabled: false
```
Comment (ichabanivan) [inline src/pages/public/sign-in/reducer.js]: initialized
errorMessage

File: src/pages/saga.js [hunk 73]
```diff
@@ -0,0 +1,87 @@
+// outsource dependencies
+import { push } from 'connected-react-router';
+import { fork, put, takeEvery, delay, call, take, cancel, cancelled } from 'redux-saga/effects';
+
+
```
Comment (ichabanivan) [inline src/pages/saga.js]:
```suggestion
```

File: src/pages/saga.js [hunk 74]
```diff
@@ -0,0 +1,87 @@
...
+    data: token
+  });
+}
+
+function * appInitialize ({ type, payload }) {
+  yield delay(500);
+  try {
+    const { accessToken } = yield call(getLocalStorage, TOKEN);
+    yield call(addAuthorizationHeader, accessToken);
+
```
Comment (ichabanivan) [inline src/pages/saga.js]:
```suggestion
```

File: src/utils/private-route.js [hunk 75]
```diff
@@ -0,0 +1,19 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { Redirect, Route } from 'react-router-dom';
+
+export default function PrivateRoute ({ component: Component, isUserAuthed, ...rest }) {
+  return (
```
Comment (ichabanivan) [inline src/utils/private-route.js:6]:
```suggestion
  return 
```

File: src/utils/private-route.js [hunk 76]
```diff
@@ -0,0 +1,19 @@
...
+    <Route
+      {...rest}
+      render={(props) =>
+        isUserAuthed ? <Component {...props} {...rest} /> : <Redirect to="/public/sign-in" />
```
Comment (ichabanivan) [inline src/utils/private-route.js:10]: Can you do it on layout level?

**Pull request merged**